### PR TITLE
Device query

### PR DESCRIPTION
**Pull request by GeorgeSheppard**
@willrogers, some initial feedback on what you think of the backend work would be helpful. Will add new tests once I know this is the right direction.

### CODE REVIEW

File: src/connection/coniql.ts [hunk 1]
```diff
@@ -317,10 +348,12 @@ export class ConiqlPlugin implements Connection {
 
   public connect(
     connectionCallback: ConnectionChangedCallback,
-    valueCallback: ValueChangedCallback
+    valueCallback: ValueChangedCallback,
+    deviceQueried: DeviceQueriedCallback
```
Comment (GeorgeSheppard) [inline src/connection/coniql.ts:352]: I dislike having to add an extra function here to handle the device query but wasn't sure there was a better way of doing it. Using the second function to handle device values as well seemed a bit opaque to me.

**Comment by willrogers** [comment]
Yes, this looks more like what I expected.

File: src/connection/coniql.ts (hunk 1 above)
Comment (willrogers) [reply src/connection/coniql.ts:352]: This looks ok to me.

**Comment by GeorgeSheppard** [comment]
@willrogers this should be the backend for devices done now hopefully

File: src/ui/hooks/utils.ts [hunk 2]
```diff
@@ -16,6 +17,16 @@ export function pvStateSelector(
   return results;
 }
 
+export function deviceSelector(device: string, state: CsState): DType {
+  return state.deviceCache[device];
+}
+
+export function deviceComparator(before: DType, after: DType): boolean {
```
Comment (willrogers) [inline src/ui/hooks/utils.ts:24]: I'm not so sure about this one. If I am correct, `deviceComparator()` will be called each time the store updates, then anything using `useDevice()` will be re-rendered each time the store updates.

I'm not certain, so I propose we leave it as it is and then we can check our understanding.

**Comment by willrogers** [comment]
This looks fine to me. Would you like to merge, or continue working on this branch?

File: src/ui/hooks/utils.ts (hunk 2 above)
Comment (GeorgeSheppard) [reply src/ui/hooks/utils.ts:24]: Ahh interesting, my understanding was that redux passes in the previous value for a specific device (found using the deviceSelector first), and then just compares the old with the new. I'll have a look into it.

**Comment by GeorgeSheppard** [comment]
Merging would be good, I'm conscious of the refactor you want to merge in so don't want this branch to get too outdated.

**Pull request merged**